runtime/src/iree/hal/drivers/vulkan/vma_allocator.cc

```diff
@@ -18,13 +19,271 @@
+    int* out_upload_idx, int* out_download_idx) {
+  int cached_idx = -1;
+  int uncached_idx = -1;
+  int visible_idx = -1;
+  for (uint32_t i = 0; i < memory_props->memoryTypeCount; ++i) {
+    VkMemoryPropertyFlags flags = memory_props->memoryTypes[i].propertyFlags;
+    if (!iree_all_bits_set(flags, include_flags) ||
+        iree_any_bit_set(flags, exclude_flags)) {
+      // Caller allows/disallows certain flags.
+      continue;
+    } else if (!iree_hal_vulkan_is_memory_type_usable(flags)) {
+      // Only want memory that is usable for storage buffers.
+      continue;
+    } else if (!iree_all_bits_set(flags, VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT)) {
+      // Must be host-visible for transfers.
+      continue;
+    }
+    if (visible_idx == -1) visible_idx = i;
+    if (iree_all_bits_set(flags, VK_MEMORY_PROPERTY_HOST_CACHED_BIT)) {
+      if (cached_idx == -1) cached_idx = i;
+    } else {
+      if (uncached_idx == -1) uncached_idx = i;
+    }
+  }
+  // Prefer uncached for uploads to enable write-through to the device.
+  *out_upload_idx = uncached_idx != -1 ? uncached_idx : visible_idx;
+  // Prefer cached for downloads to enable prefetching/read caching.
+  *out_download_idx = cached_idx != -1 ? cached_idx : visible_idx;
+}
+
+static void iree_hal_vulkan_populate_transfer_memory_types(
+    const VkPhysicalDeviceProperties* device_props,
+    const VkPhysicalDeviceMemoryProperties* memory_props,
+    iree_hal_vulkan_memory_types_t* out_types) {
+  int host_local_upload_idx = -1;
+  int host_local_download_idx = -1;
+  iree_hal_vulkan_find_transfer_memory_types(
+      device_props, memory_props, /*include_flags=*/0,
+      /*exclude_flags=*/VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT,
+      &host_local_upload_idx, &host_local_download_idx);
+  int device_local_upload_idx = -1;
+  int device_local_download_idx = -1;
+  iree_hal_vulkan_find_transfer_memory_types(
+      device_props, memory_props,
+      /*include_flags=*/VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT,
+      /*exclude_flags=*/0, &device_local_upload_idx,
+      &device_local_download_idx);
+
+  // For bulk try first to select host-local memory.
+  // In case that fails we will use device-local memory; common on integrated.
+  out_types->bulk_upload_idx = host_local_upload_idx != -1
+                                   ? host_local_upload_idx
+                                   : device_local_upload_idx;
+  out_types->bulk_download_idx = host_local_download_idx != -1
+                                     ? host_local_download_idx
+                                     : device_local_download_idx;
+
+  // Always use device-local for staging if we have it. This is BAR/page-locked
+  // memory on discrete devices but may just be the same has
+  out_types->staging_upload_idx = device_local_upload_idx != -1
+                                      ? device_local_upload_idx
+                                      : host_local_upload_idx;
```

Comment:
be the same as what?

(and what's "BAR"?)

Reply:
PCI-E magic that allows for host<->device memory mapping without host intervention (how you get a void* you can write to in device local memory); it's nifty stuff but often pretty limited: https://www.patreon.com/posts/look-at-modern-e-44371506